.gitignore

```diff
@@ -1,6 +1,6 @@
 # CakePHP specific files #
 ##########################
-/config/app.php
+/config/app_local.php
```

Comment:
`app.php` is required by bootstrap, but currently not in git and not in `.gitignore`.

Reply:
But two editable files is not OK.

It should be `app.default.php`:
- big file with full list of settings,
- secrets/credentials free, 
- tries to use environment,
- always bootstrap loaded,
- git versioned,
- production friendly.

On local machines, it should be only optional file `app.php`:
- small file with minimal set of overrides,
- no environment fallbacks (it is either environment variables or overrides file),
- has secrets and credentials,
- optionally bootstrap loaded,
- git ignored,
- copied from development friendly template `app_local.default.php` (`app.dev.php`/`app.template.php`/...?).

Reply:
Using `app.default.php` + `app.php` is no different from using `app.php` + `app_local.php`. Using the latter combination allows more flexibility for eg. instead of `app_local.php`. I for e.g. use `app_dev.php`,  `app_stagin.php`, `app_prod.php` etc. loaded based on an environment variable. Having the "local" file for each environment named separately give more clarity.

Reply:
if `app.php` is not in `gitignore` and if `secret` should be there 
so in server you should have `app_local.php` ?
if yes why you used `_local` name?

Reply:
Yes, it could be multiple templates `app_dev.php`, `app_staging.php`, `app_prod.php`.

But untracked copied files should be minimal to be easily reviewable. 

Currently `app.php` is copied from big `app.default.php`, and then it is hard keep big untracked file in sync.

It should be one big git tracked, always loaded file.

Overrides comes through environment and/or small untracked `app.php`/`app_*.php` file.

From current state, it easiest to load `app.default.php + app_local.php`.
But then there is no `app.php`, so I suggest to use `app_local.php` as template for `app.php` and load `app.default.php + app.php`

Reply:
Personally I  always load only `app.php` (copied from `app.dev.php` template):

```php
<?php
$config = require 'app.default.php';
// add/modify/unset $config keys
return $config;
```

It is easier to understand plain PHP array ops, and there is no  `Hash::merge()`.

Reply:
The template file `app.default.php` should be deleted by the installer after `app.php`. It's useless from that point onwards and can just cause confusion.

Reply:
I prefered it when the template file was named `app.php.default` as it made it clearer that it was a template file and wasn't actually loaded.

Reply:
But it would restored on next `git checkout`...

Reply:
We can always change the ignore rules as needed.

Reply:
You can't ignore file if already in git file tree.

Reply:
@garas Smartest way is two separate files in git. See why at https://sandbox.dereuromark.de/pages/best-practices#config
Then one app_local.php and done. 
Dont make it more complicated than it needs to be for beginning.
Separate files per env people can do themselves.

Reply:
@garas https://www.git-tower.com/learn/git/faq/ignore-tracked-files-in-git

Even if does get checked out for existing apps it doesn't really matter. It's just an extra file which isn't actually loaded.

But at least it will be gone for new app which start from this skeleton.

Reply:
@ADmad The page from link ignores files by removing files from git tree, because "Git can only ignore files that are untracked".

@dereuromark  So skeleton `app.default.php` should just be renamed to `app.php` here as I see then no point of copy/delete with installer. Like in sandbox. Skeleton would then have `app.php` from start. Then only app_local should be copied from template.

I just used different names:
tracked `app.default.php` + untracked `app.php` like in skeleton
You:
tracked `app.php` + untracked `app_local.php` like in sandbox.

Reply:
> `app.php` is required by bootstrap, but currently not in git and not in `.gitignore`.

It is generated from app.default.php
However I can just rename app.default.php to app.php and remove the generation part.
However TBH I would keep the generation part as it is because if someone copies in their app.php as they used to it will still work.

Reply:
> `app.php` is required by bootstrap, but currently not in git and not in `.gitignore`.

It is generated from app.default.php
However I can just rename app.default.php to app.php and remove the generation part.
However TBH I would keep the generation part as it is because if someone copies in their app.php as they used to it will still work.

Reply:
So did you 3 guys game to a conclusion that I can work on?

Reply:
>Smartest way is two separate files in git. See why at https://sandbox.dereuromark.de/pages/best-practices#config

:+1: 

for `cakephp/app` we can have only 
```
Configure::load('app', 'default', false);
Configure::load('app_custom');
```
`app.php` in git (like `app.php.default` )
`app_custom.php` in gitignore 

`app_custom` is proper name for using in both local and server and less confusing

Reply:
Noo, the only gitignore file must be named app_local as intuitivly expected.

Reply:
`app.php` (old `app.default.php`) - git tracked
`app_local.default.php` - git tracked
`app_local.php` - git ignored

Reply:
So do we have the conclusion to rename app.default.php to app.php and remove the installer parts that copy that file?

Reply:
The installer just installs app_local.default now to app_local.

Reply:
I have pushed an update, happy?